dftimewolf/lib/args_validator.py

```diff
@@ -355,6 +356,113 @@ def _ValidateOrder(self,
           f'{first} is after {second} but it should be the other way around')
 
 
+class FQDNValidator(RegexValidator):
+  """Validator for FQDNs."""
+
+  NAME = 'fqdn'
+  FQDN_REGEX = (
+      r'(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}$)')
+  # Source: https://stackoverflow.com/questions/11809631/fully-qualified-domain-name-validation#20204811  # pylint: disable=line-too-long
+  # Retrieved 2023-02-03
+
+  def ValidateSingle(self,
+                     operand: str,
+                     validator_params: Optional[Dict[str, Any]] = None) -> None:
+    """Validate an FQDN.
+
+    Args:
+      operand: The FQDN to validate.
+      validator_params: Unused for this validator.
+
+    Raises:
+      errors.RecipeArgsValidatorError: Raised if argument fails validation.
+    """
+    if not validator_params:
+      validator_params = {}
+    validator_params['regex'] = self.FQDN_REGEX
+
+    try:
+      super().ValidateSingle(operand, validator_params)
+    except errors.RecipeArgsValidatorError:
+      # Give a nicer error message than the regex failure
+      raise errors.RecipeArgsValidatorError(
+          f"'{operand}' is an invalid hostname.")
+
+
+class GRRHostValidator(FQDNValidator):
+  """Validates a GRR hostname.
+
+  Grr can accept FQDNs, or Grr client IDs, which take the form of
+  C.1facf5562db006ad.
+  """
+
+  NAME = 'grr_host'
+  GRR_REGEX = r'^C\.[0-9a-f]{16}$'
+
+  def ValidateSingle(self,
+                     operand: str,
+                     validator_params: Optional[Dict[str, Any]] = None) -> None:
+    """Validate a Grr host ID.
+
+    Args:
+      operand: The ID to validate.
+      validator_params: Unused for this validator.
+
+    Raises:
+      errors.RecipeArgsValidatorError: Raised if argument fails validation.
+    """
+    if not validator_params:
+      validator_params = {}
+
+    # Grr Host IDs can be FQDNs, test that first
+    try:
+      FQDNValidator.ValidateSingle(self, operand, validator_params)
```

Comment:
This calling convention is a bit odd - this looks like you are invoking a classmethod on FQDNValidator with a reference to self?

Reply:
`self.Method(arg)` is the same as `ClassName.Method(self, arg)` but here I needed to specify which version of `ValidateSingle` is used, from multiple potential parent classes since `self.` or `super()` seemed to pick wrong versions while writing this.